File: crates/rustacle-ipc/src/commands/terminal.rs
use serde::{Deserialize, Serialize};

/// Request to open a new terminal tab.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct OpenTabRequest {
    pub cwd: Option<String>,
}

/// Response from `open_tab`.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct OpenTabResponse {
    pub tab_id: String,
}

/// Request to split a tab into two panes.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct SplitTabRequest {
    pub tab_id: String,
    /// `"horizontal"` or `"vertical"`.
    pub direction: String,
}

/// Response from `split_tab`.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct SplitTabResponse {
    pub new_tab_id: String,
    pub split_node_id: String,
}

/// Request to resize a split divider.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ResizeSplitRequest {
    pub node_id: String,
    pub ratio: f64,
}

/// Request to reorder a tab.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct ReorderTabRequest {
    pub tab_id: String,
    pub new_index: u32,
}

/// Serializable tab info.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct TabInfo {
    pub id: String,
    pub cwd: String,
    pub title: String,
    pub alive: bool,
    pub index: u32,
    pub active: bool,
}

/// A node in the recursive split layout tree.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum SplitLayout {
    Leaf {
        tab_id: String,
    },
    Split {
        id: String,
        direction: String,
        ratio: f64,
        children: Vec<SplitLayout>,
    },
}

/// Per-tab agent context returned to the frontend.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct TabContextResponse {
    pub cwd: String,
    pub last_commands: Vec<CommandRecord>,
}

/// A single command record from tab history.
#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
pub struct CommandRecord {
    pub command: String,
    pub exit_code: i32,
    pub timestamp_ms: u64,
}

File: ui/src/state/terminal.ts
/**
 * Terminal state management: tabs, active tab, split layout.
 *
 * Uses a reducer pattern matching agent.ts conventions.
 */

export interface TabInfo {
  id: string;
  cwd: string;
  title: string;
  alive: boolean;
  index: number;
  active: boolean;
}

/** Recursive split layout tree from the backend. */
export type SplitLayout =
  | { kind: "leaf"; tab_id: string }
  | { kind: "split"; id: string; direction: string; ratio: number; children: SplitLayout[] };

export interface TerminalState {
  tabs: TabInfo[];
  activeTabId: string | null;
  layout: SplitLayout | null;
}

export const initialTerminalState: TerminalState = {
  tabs: [],
  activeTabId: null,
  layout: null,
};

export type TerminalAction =
  | { type: "SET_TABS"; tabs: TabInfo[] }
  | { type: "ADD_TAB"; tab: TabInfo }
  | { type: "CLOSE_TAB"; tabId: string }
  | { type: "SET_ACTIVE"; tabId: string }
  | { type: "UPDATE_LAYOUT"; layout: SplitLayout | null }
  | { type: "REORDER"; tabId: string; newIndex: number }
  | { type: "SET_TAB_TITLE"; tabId: string; title: string };

export function terminalReducer(state: TerminalState, action: TerminalAction): TerminalState {
  switch (action.type) {
    case "SET_TABS":
      return {
        ...state,
        tabs: action.tabs,
        activeTabId: action.tabs.find((t) => t.active)?.id ?? state.activeTabId,
      };

    case "ADD_TAB":
      return {
        ...state,
        tabs: [...state.tabs, action.tab],
        activeTabId: action.tab.id,
      };

    case "CLOSE_TAB": {
      const remaining = state.tabs.filter((t) => t.id !== action.tabId);
      const newActive =
        state.activeTabId === action.tabId
          ? remaining[remaining.length - 1]?.id ?? null
          : state.activeTabId;
      return { ...state, tabs: remaining, activeTabId: newActive };
    }

    case "SET_ACTIVE":
      return { ...state, activeTabId: action.tabId };

    case "UPDATE_LAYOUT":
      return { ...state, layout: action.layout };

    case "REORDER": {
      const tabs = [...state.tabs];
      const oldIdx = tabs.findIndex((t) => t.id === action.tabId);
      if (oldIdx === -1) return state;
      const [moved] = tabs.splice(oldIdx, 1);
      tabs.splice(action.newIndex, 0, moved);
      return { ...state, tabs: tabs.map((t, i) => ({ ...t, index: i })) };
    }

    case "SET_TAB_TITLE":
      return {
        ...state,
        tabs: state.tabs.map((t) =>
          t.id === action.tabId ? { ...t, title: action.title } : t
        ),
      };

    default:
      return state;
  }
}
